Feature/rcb 5 get public recipes

Muntean-Ionut opened the pull request:
I created the endpoint to retrieve a list with all public recipes and two filters for it: Recipe Name and Recipe Type.

Muntean-Ionut committed:
RCB-5-Create recipe controller

Muntean-Ionut committed:
RCB-5-Create recipe controller

Muntean-Ionut committed:
added RecipeService and RecipeRepository

Muntean-Ionut committed:
Merge branch 'main' into feature/RCB-5-get_public_recipes

Muntean-Ionut committed:
RCB-5-Create recipe controller

Muntean-Ionut committed:
Created CustomRecipe and CustomRecipeMapper

I im using MapStruct so i have added the required dependency as well.

Muntean-Ionut committed:
Merge branch 'main' of https://github.com/EndavaCollab/bvdapprenticeship2024-recipe-box into feature/RCB-5-get_public_recipes

Muntean-Ionut committed:
Removed CustomRecipeMapperImpl

CustomRecipeMapperImpl is not needed anymore because MapStruct already does the implementation.

en-madalina-marales commented on recipe-box-backend/pom.xml:
```diff
@@ -47,8 +48,24 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.17.2</version>
```
As a recommendation, put all the versions in properties section

en-madalina-marales commented on recipe-box-backend/pom.xml:
```diff
@@ -47,8 +48,24 @@
+        <dependency>
+            <groupId>org.mapstruct</groupId>
+            <artifactId>mapstruct-processor</artifactId>
+            <version>1.6.0</version>
```
Same recommendation as above

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/service/RecipeService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public List<Recipe> getAllPublicRecipes() {
+        List<Recipe> recipes = List.copyOf(recipeRepository.findAll()).stream()
```
Local variables should not be declared and then immediately returned or thrown, this adds complexity in reading the code. Please return directly.

en-madalina-marales replied:
List.copyOf() is adding unnecessary overhead of creating an immutable copy of the list. 
The primary reason to use List.copyOf() would be if you want to ensure that the list of recipes returned by findAll() is immutable before streaming over it. However, in this context, you immediately create a stream and perform operations on it, followed by collecting the results into a new list.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/service/RecipeService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public List<Recipe> getAllPublicRecipesByName(String recipeName) {
+        List<Recipe> recipes = recipeRepository.findAll().stream()
```
Local variables should not be declared and then immediately returned or thrown, this adds complexity in reading the code. Please return directly.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/service/RecipeService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public List<Recipe> getAllPublicRecipesByType(MealType mealType) {
+        List<Recipe> recipes = recipeRepository.findAll().stream()
```
Local variables should not be declared and then immediately returned or thrown, this adds complexity in reading the code. Please return directly.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+public class RecipeController {
+
+    public final RecipeService recipeService;
+    private final CustomRecipeMapper customRecipeMapperMapper = Mappers.getMapper(CustomRecipeMapper.class);
```
The CustomRecipeMapper is initialized directly in the field declaration. While this is not inherently wrong, it can make testing difficult if you ever need to mock or replace the mapper. Consider injecting CustomRecipeMapper via the constructor, similar to RecipeService.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @GetMapping("/publicRecipes/{name}")
+    public ResponseEntity<?> getAllPublicRecipesByName(@PathVariable(name = "recipeName") String recipeName) {
```
The return type ResponseEntity<?> is used across all methods. While this is acceptable, specifying a more concrete type would improve type safety and readability.
If the return type is known, use ResponseEntity<List<CustomRecipeDto>> or a specific DTO class.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+    @GetMapping("/publicRecipes/{name}")
+    public ResponseEntity<?> getAllPublicRecipesByName(@PathVariable(name = "recipeName") String recipeName) {
+        if(recipeName == null)
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(String.format("The recipe name entered is null"));
```
Consider adding logging your controller and service methods, especially when handling exceptions or invalid input, to improve traceability. Example: private static final Logger logger = LoggerFactory.getLogger(RecipeController.class);
and before returning the bad request, add logger.warn("Received a null recipe name");

en-madalina-marales replied:
Format strings should be used correctly, format string is a string that contains placeholders, usually represented by special characters such as "%s" or "{}", depending on the technology in use. These placeholders are replaced by values when the string is printed or logged.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+@RequestMapping(path = "/recipe_box", produces = "application/json")
+public class RecipeController {
+
+    public final RecipeService recipeService;
```
this one can be private and moved into service class

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+    @GetMapping("/publicRecipes/{mealType}")
+    public ResponseEntity<?> getAllPublicRecipesByType(@PathVariable(name = "mealType") MealType mealType) {
+        if(mealType == null)
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(String.format("The recipe type entered is null"));
```
Format strings should be used correctly, format string is a string that contains placeholders, usually represented by special characters such as "%s" or "{}", depending on the technology in use. These placeholders are replaced by values when the string is printed or logged.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+    public ResponseEntity<?> getAllPublicRecipesByType(@PathVariable(name = "mealType") MealType mealType) {
+        if(mealType == null)
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(String.format("The recipe type entered is null"));
+        return ResponseEntity.ok(customRecipeMapperMapper.map(recipeService.getAllPublicRecipesByType(mealType)));
```
the mapping part must be included into service class. and here we just need to return the response of service method

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+    public ResponseEntity<?> getAllPublicRecipesByName(@PathVariable(name = "recipeName") String recipeName) {
+        if(recipeName == null)
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(String.format("The recipe name entered is null"));
+        return ResponseEntity.ok(customRecipeMapperMapper.map(recipeService.getAllPublicRecipesByName(recipeName)));
```
the mapping part must be included into service class. and here we just need to return the response of service method

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+
+
+@RestController
+@RequestMapping(path = "/recipe_box", produces = "application/json")
```
receipe_box si the name of the entire application, in this case we need to reflect the controller, so I recommand to use only "/recipes" inside RequestMapping.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/service/RecipeService.java:
```diff
@@ -0,0 +1,55 @@
+    public List<Recipe> getAllPublicRecipesByName(String recipeName) {
+        List<Recipe> recipes = recipeRepository.findAll().stream()
+                .filter(RecipeService::isPublic)
+                .filter(r -> Objects.equals(r.getName(), recipeName))
```
for comparing 2 strings we shouldn't use method from Objects. You can check directly **.filter(r -> r.getName().equals(recipeName))**

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+
+    @GetMapping("/publicRecipes")
+    public ResponseEntity<?> getAllPublicRecipes() throws IOException {
+        return ResponseEntity.ok(customRecipeMapperMapper.map(recipeService.getAllPublicRecipes()));
```
the mapping part must be included into service class. and here we just need to return the response of service method

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+        this.recipeService = recipeService;
+    }
+
+    @GetMapping("/publicRecipes")
```
My recommendation is to change from "recipe_box/publicRecipes" to "recipes/". Is a best practice to use: Clear and Consistent Resource Naming. Use nouns that accurately reflect the resources your API manages (e.g., /recipes, /users).
Because the controller is already mapped with /recipes, on this method you will need to map to a "/". And for the full picture, when the endpoint with get private recipes will be implemented, because we will need to filter recipes by user Id, we will have an endpoint like "/recipes/users/{id}".

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+        return ResponseEntity.ok(customRecipeMapperMapper.map(recipeService.getAllPublicRecipes()));
+    }
+
+    @GetMapping("/publicRecipes/{name}")
```
I recommand to follow the recommendations from RESTful Design Principles, and to use /recipes/search?name=pizza. So in this case the GetMapping("/search") and with a RequestParam "name".

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,51 @@
+        return ResponseEntity.ok(customRecipeMapperMapper.map(recipeService.getAllPublicRecipesByName(recipeName)));
+    }
+
+    @GetMapping("/publicRecipes/{mealType}")
```
Do we really need this method? In the descript I saw mentioned the recipeType and not the meal type. I think that we can get rid of this method. You can ask in daily if we really need an endpoint that is getting the recipes by meal type. Please ask this, because I will not be able to join daily. If not, ask on the team channel.

Muntean-Ionut committed:
Updated Recipe Endpoint

Ionel14 commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @GetMapping("/")
+    public ResponseEntity<List<CustomRecipe>> getAllPublicRecipes(@RequestParam(required = false) Long id) {
+        return ResponseEntity.ok(recipeService.getAllPublicRecipes());
+    }
```
Here you did nothing with the id parameter. I guess that is for filtering by meal type, also, a suggestive name will be good.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/service/RecipeService.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.stream.Collectors;
+
+@Service
+public class RecipeService {
```
Recipe Service should be an interface. And you need to create a class RecipeServiceImpl that implements the interface. And from the controller you will call the interface. The reason to user interfaces is to use one of the OOP concepts: abstraction and decoupling.

en-madalina-marales replied:
Recipe Service should be an interface. And you need to create a class RecipeServiceImpl that implements the interface. And from the controller you will call the interface. The reason to user interfaces is to use one of the OOP concepts: abstraction and decoupling.

en-madalina-marales reviewed (changes requested):
Please use the abstraction on service layer. By adding an interface that will be implemented by class recipeServiceImpl

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/controller/RecipeController.java:
```diff
@@ -1,51 +1,36 @@
-    public ResponseEntity<?> getAllPublicRecipes() throws IOException {
-        return ResponseEntity.ok(customRecipeMapperMapper.map(recipeService.getAllPublicRecipes()));
+    @GetMapping("/")
+    public ResponseEntity<List<CustomRecipe>> getAllPublicRecipes(@RequestParam(required = false) Long id) {
```
The requested param should be mealType. Because we need to filter all public recipes by mealType.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/response/CustomRecipe.java:
```diff
@@ -0,0 +1,41 @@
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class CustomRecipe  {
```
A best practice in naming of Objects that we send on the response is to be named as RecipeDTO, DTO = Data Transfer Object.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/response/CustomRecipeMapper.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.List;
+
+@Mapper
+public interface CustomRecipeMapper {
```
Inside the mapper we can have multiple mappings between different type of wrappers created for Recipes. This is why I prefer to be names as RecipeMapper

Muntean-Ionut committed:
Updated Recipe Endpoint

Created interface RecipeServiceImpl
Renamed CustomRecipe to RecipeDTO
Renamed CustomRecipeMapper to RecipeMapper

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/service/RecipeServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.stream.Collectors;
+
+@Service
+public class RecipeServiceImpl implements RecipeService{
```
all classes that will implement the service interfaces should be existing into directory /service/impl. This will make the code repo structure to be easily to follow.

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/response/RecipeMapper.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.List;
+
+@Mapper
+public interface RecipeMapper {
```
all the mapper classes that will be creating should stay into directory /mapper. This directory in on the same level as /controller

en-madalina-marales commented on recipe-box-backend/src/main/java/com/endava/recipebox/response/RecipeDTO.java:
```diff
@@ -0,0 +1,41 @@
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class RecipeDTO {
```
The standard for directory that holds all the DTO objects is /dto. You can rename from /response to /dto

Muntean-Ionut committed:
Changed the structure of the project

Made "RecipeServiceImpl" package and moved the corresponding class there
Made "mapper" package and moved the corresponding interface there
Renamed "response" package to "dto"
